graph/interfaces.go

```diff
@@ -88,3 +90,191 @@ type ChannelGraphSource interface {
+		*models.ChannelEdgePolicy, *models.ChannelEdgePolicy, error)
+
+	// AddLightningNode adds a vertex/node to the graph database. If the
+	// node is not in the database from before, this will add a new,
+	// unconnected one to the graph. If it is present from before, this will
+	// update that node's information. Note that this method is expected to
+	// only be called to update an already present node from a node
+	// announcement, or to insert a node found in a channel update.
+	AddLightningNode(node *channeldb.LightningNode,
+		op ...batch.SchedulerOption) error
+
+	// AddChannelEdge adds a new (undirected, blank) edge to the graph
+	// database. An undirected edge from the two target nodes are created.
+	// The information stored denotes the static attributes of the channel,
+	// such as the channelID, the keys involved in creation of the channel,
+	// and the set of features that the channel supports. The chanPoint and
+	// chanID are used to uniquely identify the edge globally within the
+	// database.
+	AddChannelEdge(edge *models.ChannelEdgeInfo,
+		op ...batch.SchedulerOption) error
+
+	// MarkEdgeZombie attempts to mark a channel identified by its channel
+	// ID as a zombie. This method is used on an ad-hoc basis, when channels
+	// need to be marked as zombies outside the normal pruning cycle.
+	MarkEdgeZombie(chanID uint64, pubKey1, pubKey2 [33]byte) error
+
+	// UpdateEdgePolicy updates the edge routing policy for a single
+	// directed edge within the database for the referenced channel. The
+	// `flags` attribute within the ChannelEdgePolicy determines which of
+	// the directed edges are being updated. If the flag is 1, then the
+	// first node's information is being updated, otherwise it's the second
+	// node's information. The node ordering is determined by the
+	// lexicographical ordering of the identity public keys of the nodes on
+	// either side of the channel.
+	UpdateEdgePolicy(edge *models.ChannelEdgePolicy,
+		op ...batch.SchedulerOption) error
+
+	// HasLightningNode determines if the graph has a vertex identified by
+	// the target node identity public key. If the node exists in the
+	// database, a timestamp of when the data for the node was lasted
+	// updated is returned along with a true boolean. Otherwise, an empty
+	// time.Time is returned with a false boolean.
+	HasLightningNode(nodePub [33]byte) (time.Time, bool, error)
+
+	// FetchLightningNode attempts to look up a target node by its identity
+	// public key. If the node isn't found in the database, then
+	// ErrGraphNodeNotFound is returned.
+	FetchLightningNode(nodePub route.Vertex) (*channeldb.LightningNode,
+		error)
+
+	// ForEachNode iterates through all the stored vertices/nodes in the
+	// graph, executing the passed callback with each node encountered. If
+	// the callback returns an error, then the transaction is aborted and
+	// the iteration stops early.
+	ForEachNode(cb func(kvdb.RTx, *channeldb.LightningNode) error) error
```

Comment:
Still going through the PR, so ignore if resolved later. Just wondering if these last two references to `kvdb` here and below will be replaced/refactored later to make this fully independent of an actual local database implementation?

Reply:
Sure, sounds good to me :+1: